package.json

```diff
@@ -36,6 +36,7 @@
     "ant-design-pro": "^2.1.1",
     "antd": "^3.16.1",
     "classnames": "^2.2.5",
+    "dotenv": "^8.2.0",
```

Comment:
`dotenv` reads the contents of the `.env` file when the Node.js process boots up. It creates an `env` object as a property of the global `process` object. This workflow doesn't involve network requests and is the reason we can reference configuration at the earliest point in dashboard initialization.

Reply:
But is the `.env` file built into the index.html page that is loaded?  Or the `<script ...>` references in the `index.html` back on the server?  How does Node.js running in the browser get the contents of the `.env`?  Or are you talking about a Node.js process on the server that reads this?

Reply:
The bundled dashboard code is configured to read the `.env` file at the root of the deployed dashboard directory. This is done by adding `require('dotenv').config();` to the UmiJS config which is then bundled into the `umi.js` file referenced in `index.html`. We don't require a Node.js process to run on a target server in order to serve the `.env` file.

Reply:
> The bundled dashboard code is configured to read the `.env` file at the root of the deployed dashboard directory. This is done by adding `require('dotenv').config();` to the UmiJS config which is then bundled into the `umi.js` file referenced in `index.html`. We don't require a Node.js process to run on a target server in order to serve the `.env` file.

So then the `umi.js` file is _built_ with the contents of the `.env` file, and then the `umi.js` file is deployed to the target server?

Do we all agree that the principle of creating a deployable `umi.js` file once and deploying it to multiple environments is one we want to adhere to as a project?

If we all agree, great.

If we don't agree with that principle, we need to discuss that first because we need a guide for how we handle builds and deployment.

That principle is at the heart of containers (build the container once, deploy it via different orchestrators [podman, kubernetes, others]), of Ansible (roles and playbooks are applied to various inventory files with different contents), and CI/CD pipelines (build once, test the same object without changes through the pipeline until deployed).

Reply:
> > The bundled dashboard code is configured to read the `.env` file at the root of the deployed dashboard directory. This is done by adding `require('dotenv').config();` to the UmiJS config which is then bundled into the `umi.js` file referenced in `index.html`. We don't require a Node.js process to run on a target server in order to serve the `.env` file.
> 
> So then the `umi.js` file is _built_ with the contents of the `.env` file, and then the `umi.js` file is deployed to the target server?

Hmm; good question, but there is a `/var/www/html/dashboard/.env` on the staging server, and my impression was that `dotenv` is somehow sucking that in. I don't really see how that could happen with a browser and the static dashboard Javascript on the server unless it's doing an HTML GET to retrieve it just like it does the CSS.

> Do we all agree that the principle of creating a deployable `umi.js` file once and deploying it to multiple environments is one we want to adhere to as a project?

Yes, very strongly!

Reply:
I agree with the "build once, deploy anywhere" paradigm. The aim of this PR is to bridge the gap between deploying to multiple environments with one binary while also allowing endpoint config to be referenced at dashboard initialization. The discussion in this issue captures why this is an uncommon pattern in React environments and generally requires a "hacked" workaround: https://github.com/facebook/create-react-app/issues/2353

I've reviewed suggested workarounds and come up with a solution to configure endpoints at dashboard runtime so we avoid bundling endpoints into the dashboard binary. We do this by referencing ` <script src="<%= context.config.publicPath +'endpoints.js'%>"></script>` in the dashboard `index.html` template and fetching `endpoints.js` with a `GET` request. `endpoints.js` is included in `dist/` as a standalone file and can be modified after deployment.

I've deployed the dashboard to staging using this model. Feel free to inspect the Network tab in dev tools to see the `endpoints.js` file being queried. Additionally, you can modify `endpoints.js` and refresh the dashboard to try different API endpoints.